PushSharp.Apple/AppleHttpPushChannel.cs

```diff
@@ -0,0 +1,210 @@
+
+                    // Should we delete the device token when: HttpStatusCode.BadRequest - BadDeviceToken - The specified device token is invalid.Verify that the request contains a valid token and that the token matches the environment.
+                    // special case to call delete device token, means that device token it no longer available on apple side
+                    if (response.StatusCode == HttpStatusCode.Gone)
+                    {
+                        if (callback != null)
+                        {
+                            SendNotificationResult result;
+                            if (_appleSettings.EnableDeleteTokenOn410Response)
+                            {
+                                result = new SendNotificationResult(notification, false, new Exception("Device token no longer available on APNs."))
+                                {
+                                    IsSubscriptionExpired = true,
+                                    OldSubscriptionId = appleNotification.DeviceToken
+                                };
+                            }
+                            else
+                            {
+                                result = new SendNotificationResult(notification, false);
+                            }
+                            callback(this, result);
+                            return;
+                        }
+                    }
+
+                    var retryNotification = true;
+                    // this response are non recoverable, do not retry the notification
+                    switch (response.StatusCode)
+                    {
+                        case HttpStatusCode.Forbidden:
+                        case HttpStatusCode.BadRequest:
+                        case HttpStatusCode.NotFound:
+                        case HttpStatusCode.MethodNotAllowed:
+                        case HttpStatusCode.RequestEntityTooLarge:
+                            retryNotification = false;
+                            break;
+                        default:
+                            break;
+                    }
+
+                    if (callback != null)
+                    {
+                        callback(this, new SendNotificationResult(notification, retryNotification, new Exception("Error during APNS Send.")));
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                Log.Error("Exception during APNS Send with channel {2}: {0} -> {1}",
+                    appleNotification.Identifier,
+                    ex,
+                    _channelInstanceId);
+
+                if (callback != null)
+                {
+                    // will requeue the notification
+                    callback(this, new SendNotificationResult(notification, true, ex));
+                }
+            }
+        }
+
+        private string GetJwtToken()
+        {
+            lock (_lock)
```

Comment:
don't lock here - you should only lock once you have determined that a new jwt is required.  Once you enter the lock, immediately re-check the jwt expiration (since another thread may have updated the token while current thread was waiting to enter the lock).